PR: Add results in batches and limit results on Find pane to a maximum

goanpeca opened the pull request:
<!--- Make sure to read the Contributing Guidelines:                   --->
<!--- https://github.com/spyder-ide/spyder/blob/master/CONTRIBUTING.md --->
<!--- and follow PEP 8, PEP 257 and Spyder's code style:               --->
<!--- https://github.com/spyder-ide/spyder/wiki/Dev:-Coding-Style      --->

## Description of Changes


* [x] Wrote at least one-line docstrings (for any new functions)
* [x] Added unit test(s) covering the changes (if testable)
<!--- Remember that an image/animation is worth a thousand words! --->
* [x] Included a screenshot or animation (if affecting the UI, see [Licecap](https://www.cockos.com/licecap/))

![finder](https://user-images.githubusercontent.com/3627835/68431181-93c72e00-017f-11ea-9042-7743e00cd8a3.gif)



<!--- Explain what you've done and why --->

- The find in files now adds the files in batches that increase over time in power of 2
2^1, 2^2 etc.... up to a max of 9 (512 items per load batch)
- The maximum is now configurable, default is 1000
- The stop button and search button now act as "one single button"
- Enter now toggles between searching and stoping (instead of always restart a search.
- The regex section ins hidden by default



### Issue(s) Resolved

<!--- List the issue(s) below, in the form "Fixes #1234"; one per line --->

Fixes #10264


### Affirmation

By submitting this Pull Request or typing my (user)name below,
I affirm the [Developer Certificate of Origin](https://developercertificate.org)
with respect to all commits and content included in this PR,
and understand I am releasing the same under Spyder's MIT (Expat) license.

<!--- TYPE YOUR USER/NAME AFTER THE FOLLOWING: --->
I certify the above statement is true and correct: @goanpeca 

<!--- Thanks for your help making Spyder better for everyone! --->

pep8speaks commented:
Hello @goanpeca! Thanks for opening this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`spyder/plugins/findinfiles/widgets.py`](https://github.com/spyder-ide/spyder/blob/6fa6f0a9637f8dbbcf9ac9261bda9862d0aa6819/spyder/plugins/findinfiles/widgets.py):

> [Line 214:80](https://github.com/spyder-ide/spyder/blob/6fa6f0a9637f8dbbcf9ac9261bda9862d0aa6819/spyder/plugins/findinfiles/widgets.py#L214): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (94 > 79 characters)
> [Line 242:80](https://github.com/spyder-ide/spyder/blob/6fa6f0a9637f8dbbcf9ac9261bda9862d0aa6819/spyder/plugins/findinfiles/widgets.py#L242): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (81 > 79 characters)
> [Line 245:80](https://github.com/spyder-ide/spyder/blob/6fa6f0a9637f8dbbcf9ac9261bda9862d0aa6819/spyder/plugins/findinfiles/widgets.py#L245): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (94 > 79 characters)
> [Line 258:80](https://github.com/spyder-ide/spyder/blob/6fa6f0a9637f8dbbcf9ac9261bda9862d0aa6819/spyder/plugins/findinfiles/widgets.py#L258): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (80 > 79 characters)
> [Line 944:41](https://github.com/spyder-ide/spyder/blob/6fa6f0a9637f8dbbcf9ac9261bda9862d0aa6819/spyder/plugins/findinfiles/widgets.py#L944): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 965:80](https://github.com/spyder-ide/spyder/blob/6fa6f0a9637f8dbbcf9ac9261bda9862d0aa6819/spyder/plugins/findinfiles/widgets.py#L965): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (81 > 79 characters)
> [Line 1061:80](https://github.com/spyder-ide/spyder/blob/6fa6f0a9637f8dbbcf9ac9261bda9862d0aa6819/spyder/plugins/findinfiles/widgets.py#L1061): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (87 > 79 characters)

goanpeca commented:
@impact27 what do you think a reasonable limit is?

@ccordoba12 thoughts on this?

ccordoba12 commented:
> a maximum limit to results after which a message is displayed and the search is stopped

If the changes you've made here prevent Spyder to freeze, why do we need to stop searching at some point and show a message?

goanpeca commented:
> If the changes you've made here prevent Spyder to freeze,

They do not prevent it, it just makes it less terrible. But as the list keeps growing and growing it will still become unusable so we really need to set a hard limit.

impact27 commented:
> > If the changes you've made here prevent Spyder to freeze,
> 
> It prevents the deadlock freeze, but as the list keeps growing Spyder is not very responsive since every update of the list (even in batch) blocks the GUI (as you can see on the animation, the spinner stops for a moment every time the batch is loaded into the tree).
> 
> Thoughts @impact27 could you test this?

Is it possible to do part of the processing in a thread? I guess only GUI operations need to be done in the main thread. What part of the processing actually takes time?

impact27 commented:
> @impact27 what do you think a reasonable limit is?

I am not sure I would want to go through more than 1000 results. This could be a setting that is initially at 1000.

impact27 commented:
For reference, searchin for `in` took 15 min:
 - 10.5 min in `spyder/plugins/findinfiles/widgets.py::ItemDelegate::sizeHint`:
    - 6 min in `idealWidth`
    - 3.5 min in `setHtml`
 - 3 min in `spyder/plugins/findinfiles/widgets.py::FindInFilesWidget::search_complete`:
    - 3 min in `expandAll`
 - 1.5 min in `spyder/plugins/findinfiles/widgets.py:: ResultsBrowser::append_result`

goanpeca commented:
@impact27 

>For reference, searching for in took 15 min:

Are these values with this PR? or with master?

impact27 commented:
Oups you are right somehow I didn’t checkout the correct branch. So these values are for the master.

goanpeca commented:
>Is it possible to do part of the processing in a thread? I guess only GUI operations need to be done in the main thread. What part of the processing actually takes time?

The search is donde in a thread, I will try to move everything except the actual loading of the tree and see how it looks.

goanpeca commented:
Could you provide some numbers for this branch :-) ?

impact27 commented:
I’ll try to do this tomorrow

impact27 commented:
The interface is unresposnsive for ~10s (which is much better) I ran 3 searches:
 - 18 sec in `spyder/plugins/findinfiles/widgets.py:: ResultsBrowser::append_result`:
    - 10 sec in `__init__::__repr__::get_font` (in spyder/config/gui.py)
 - 5 sec in `paintEvent` (spyder/plugins/editor/widgets/codeeditor.py)

impact27 commented:
Maybe `get_font` could be cached?

goanpeca commented:
@impact27 I incorporated your suggestion and added some extra stuff. Please see the Animation on the description.

If you could test this, it would be great!

ccordoba12 commented:
> Please see the Animation on the description.

Things look pretty slick now @goanpeca! Thanks a lot for working on this!

goanpeca commented:
I still need to add 1 test , working on it now.

goanpeca commented:
This one is ready @ccordoba12 and @impact27 for review and final testing!

impact27 commented:
This totally solves the issue for me.

ccordoba12 commented on spyder/plugins/findinfiles/plugin.py:
```diff
@@ -53,24 +53,30 @@ def __init__(self, parent=None):
         case_sensitive = self.get_option('case_sensitive')
         path_history = self.get_option('path_history', [])
         search_in_index = self.get_option('search_in_index', default=0)
+        max_results = self.get_option('max_results', default=1000)
```
Please add this option to our DEFAULTS dict in `config/main.py` and remove the `default=1000` from here.

goanpeca replied:
Ok

ccordoba12 commented on spyder/plugins/findinfiles/widgets.py:
```diff
@@ -181,40 +196,157 @@ def find_string_in_file(self, fname):
+
         self.completed = True
 
+    def process_results(self):
```
Please add a docstring to this method to understand what it does.

goanpeca replied:
Done

ccordoba12 commented on spyder/plugins/findinfiles/widgets.py:
```diff
@@ -181,40 +196,157 @@ def find_string_in_file(self, fname):
+        self.partial_results = []
+        self.sig_line_match.emit(items, title)
+
+    def truncate_result(self, line, start, end):
```
Please add a docstring to this too.

goanpeca replied:
Done

ccordoba12 commented on spyder/config/main.py:
```diff
@@ -516,6 +516,10 @@
+            ('find_in_files',
+             {
+              'max_results': 1000,
+              }),
```
`find_in_files` already has its own section, so please add this entry there.

Also, please change the `more_options` of `find_in_files` to False, to hide the exclusion options by default from our config.

goanpeca replied:
Fixed (I actually had added the option in the wrong place)

ccordoba12 commented on spyder/config/main.py:
```diff
@@ -288,8 +288,9 @@
               'search_text': [''],
               'search_text_samples': [TASKS_PATTERN],
-              'more_options': True,
+              'more_options': False,
```
You forgot to bump CONF_VERSION to account for this change.

goanpeca replied:
🤦‍♂

goanpeca replied:
Fixed

jnsebgosselin commented:
Thanks a lot @goanpeca for working on this! Find in files is one of my favorite feature in Spyder actually, so I'm glad it got an improvement.

Would you guys mind if I open a PR to reduce a little bit the vertical space between the items, because I think they are a little bit too spaced now after this PR. 

![image](https://user-images.githubusercontent.com/10170372/68512612-e0c20780-0246-11ea-8340-c6ae1b98c5eb.png)

goanpeca commented:
>Would you guys mind if I open a PR to reduce a little bit the vertical space between the items, because I think they are a little bit too spaced now after this PR.

@jnsebgosselin that had to be done like that for performance reasons. We can set the same height for all items which avoids recalculating heights everytime. For big searches the time spend  on sizeHint is too much.

jnsebgosselin commented:
Cool! I hope you find a workaround for this, thanks.

goanpeca commented:
>Cool! I hope you find a workaround for this, thanks.

@jnsebgosselin so I tried and this method is a bottleneck, so Spyder becomes completely unresponsive if the amount of results is big (this needs a better definition).

So the workaround is still calling for every item even if I cache the heigh values with the first entries.

The only workaround would be to us my "hack" for maximum results bellow... 1000, but for anything above that value, go back to fixed heights.

Thoughts @ccordoba12 ?

jnsebgosselin commented:
Ok thanks for looking into it. I will try to take a look at it in case I can find a solution.

ccordoba12 commented:
> The only workaround would be to us my "hack" for maximum results bellow... 1000, but for anything above that value, go back to fixed heights.

Sounds good to me.

jnsebgosselin commented:
> > The only workaround would be to us my "hack" for maximum results bellow... 1000, but for anything above that value, go back to fixed heights.
> 
> Sounds good to me.

Ok, I've looked into this a little bit and yeah... considering the amount of data involved sometimes, I think the only solution to fix this would be to use a QTreeView instead of a QTreeWidget, but that may require a lot of work...

So like @ccordoba12, I agree with the workaround you are proposing too.

jnsebgosselin commented:
What if we go with a GUI design similar to the file switcher? That would solve the variable lines height issue while not altering the performances. Also, this doesn't look bad at all...

![image](https://user-images.githubusercontent.com/10170372/68521758-3f50ab00-0272-11ea-953d-e5d19ce95d14.png)

goanpeca commented:
Fine by me. Thoughts @ccordoba12 ?

ccordoba12 commented:
I like it too. Please go ahead with that idea @jnsebgosselin!
